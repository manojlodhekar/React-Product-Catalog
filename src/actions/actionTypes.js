export const CREATE_COURSE = 'CREATE_COURSE';
export const SEARCH_PRODUCT = 'SEARCH_PRODUCT';
export const GET_PRODUCT_ITEM = 'GET_PRODUCT_ITEM';
export const PRODUCT_API_ERROR = 'PRODUCT_API_ERROR';
export const SEARCHED_PRODUCT_TEXT = 'SEARCHED_PRODUCT_TEXT';
export const SET_RECCOMENDATION_PRODUCTS = 'SET_RECCOMENDATION_PRODUCTS'