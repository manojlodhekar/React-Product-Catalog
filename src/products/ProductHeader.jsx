import React from 'react';
import { Image, Grid, Divider } from 'semantic-ui-react'
import productConst from '../const/productConst';

const styles = {
  
    header:{
        paddingLeft: 10,
        textAlign: 'left',
        paddingBottom: 10,
        borderBottom: '1px solid grey'
    },
    headerTxt:{
        paddingTop: 10,
        textAlign: 'left',
        fontSize: 24,
        fontWeight: 'bold',
        lineHeight: 1.2
    },
    logoCls:{
        maxWidth: 60, 
        float: 'right', 
        textAlign: 'right'
    },
    subHeaderText:{
        color: '#0071ce',
        textAlign: 'left',
        fontSize: 18
    }

}

const ProductHeader = (props) =>{

    return(
        <div>
            <Grid >  
                <Grid.Row style={{paddingLeft: 20}}>
                    <Grid.Column width={3} style={{float: 'right'}}>
                        <Image src='/assets/Walmart_Spark.png' style={styles.logoCls} />
                    </Grid.Column>
                    <Grid.Column width={13}>
                        <div style={styles.headerTxt}>
                                {productConst.HEADER_TEXT}
                        </div>
                        <div style={styles.subHeaderText}>
                                {props.subHeaderText}
                        </div>
                    </Grid.Column>
                </Grid.Row>
            </Grid>
            <Divider />

           {/* <Header as='h2' style={styles.header}>
                    <Image src='/assets/Walmart_Spark.png' width="60"/> 
                    <Header.Content style={styles.headerTxt}>
                    {productConst.HEADER_TEXT}
                    <Header.Subheader>{props.subHeaderText}</Header.Subheader>
                    </Header.Content>
                </Header> */}
        </div>
    )
}
export default ProductHeader;