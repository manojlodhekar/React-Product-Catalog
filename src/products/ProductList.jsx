import React from 'react';
import { Item,} from 'semantic-ui-react'
import { Link } from 'react-router-dom';
import productConst from '../const/productConst';
import { htmlDecode, getStarRating} from '../utils/ProductUtils';

const ProductList = (props) =>{
debugger;
const productSelected = (item) =>{
    console.log(item.name);
    props.productSelection(item);
}
const renderProductList = () =>{

    let productDetail =  props.items.map((item) => {
        let longDesc = htmlDecode(item.longDescription);
        return(
            
            <div style={{padding: 15, borderBottom: '1px solid grey'}} key={item.itemId}>
                <Item.Group divided className="listContainer">
                    <Item>
                       <Link to={props.linkUrl + item.itemId } > <Item.Image size='small' src={item.mediumImage} 
                        onClick={ (e) => productSelected(item)}/> 
                        </Link>

                        <Item.Content className="listContent">
                        <Link to={props.linkUrl + item.itemId } >
                        <Item.Header className="productHeader"
                            onClick={ (e) => productSelected(item)}>{item.name}</Item.Header>
                            </Link>
                            <Item.Meta className="productHeader">Price : $ {item.salePrice}</Item.Meta>
                            <Item.Meta className="moreInfo"> {getStarRating(item.customerRating)} {item.numReviews}</Item.Meta>
                            <Item.Description>
                            <div dangerouslySetInnerHTML={{__html: longDesc}}/>
                            </Item.Description>
                        </Item.Content>
                    </Item>
                </Item.Group>

                {/* <Grid >
                    
                    <Grid.Row>
                    <Grid.Column width={3}>
                        <Link to={"/product/" + item.itemId } > <Image size='small' src={item.thumbnailImage} wrapped 
                        onClick={ (e) => productSelected(item)}/> 
                        </Link>
                    </Grid.Column>
                    <Grid.Column width={10}>
                    <Link to={"/product/" + item.itemId } >
                        <div onClick={ (e) => productSelected(item)}>{item.name} </div>
                        <span>{item.shortDescription}</span>
                    </Link>
                    </Grid.Column>
                    <Grid.Column width={3}>
                        <div>{item.salePrice}</div>
                        <div> {item.offerType}</div>
                    </Grid.Column>
                    </Grid.Row>
                </Grid> */}


            </div>
        )
    });
    return productDetail;
}
    return(
        <div>
            {props.items && props.items.length > 0 ? 
            renderProductList(): productConst.SEARCHED_PRODUCT_MESSAGE}
        </div>
    )
}


export default ProductList;