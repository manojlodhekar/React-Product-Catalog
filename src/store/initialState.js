export default {
    products: {
        productSearchText: '',
        productList:[],
        productReccomendationList: [],
        selectedProductDetail: {},
        productApiError: false
    }
}