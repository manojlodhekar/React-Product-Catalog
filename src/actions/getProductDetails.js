import * as type from '../actions/actionTypes';
import productConst from '../const/productConst';
import axios from 'axios';
import { push, } from 'react-router-redux'

export const getProductDetails = (itemId) => (dispatch) =>{
    

  let productLookupApiURL = productConst.PRODUCT_LOOKUP_API_URL +'/' + itemId + '?format=json&apiKey=' + productConst.PRODUCT_API_KEY; ;  
   

   const getProductItemSuccess = (response) => {
    return {
      type: type.GET_PRODUCT_ITEM,
      payload: response
    }
  }
  
  const getProductItemError = (err) => {
    return {
      type: type.PRODUCT_API_ERROR,
      payload: true
    }
  }


  return axios.get(productLookupApiURL)
    .then((response) => {
      dispatch(getProductItemSuccess(response.data))
      dispatch(push(`/product/${itemId}`))
    })
    .catch((err) => {
      dispatch(getProductItemError(err))
    })
    
};