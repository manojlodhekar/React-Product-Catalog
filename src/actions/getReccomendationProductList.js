import * as type from '../actions/actionTypes';
import productConst from '../const/productConst';
import axios from 'axios';
import { push, } from 'react-router-redux'

export function getReccomendationProductList(itemId){
    

  let productLookupApiURL = '';  
  let productReccomendationApiUrl = productConst.PRODUCT_RECCOMENDATION_API_URL + productConst.PRODUCT_API_KEY + '&itemId=' + itemId; 

  //http://api.walmartlabs.com/v1/nbp?apiKey=afffhq7phcwhc28kb45skp9d&itemId=118505871

   const getProductReccomendationItemsSuccess = (response) => {
    return {
      type: type.SET_RECCOMENDATION_PRODUCTS,
      payload: response.items
    }
  }
  
  const getProductError = (err) => {
    return {
      type: type.PRODUCT_API_ERROR,
      payload: true
    }
  }

  const getProductReccomendation = (url) => {
    return function(dispatch) {
      axios.get(url)
        .then((response) => {
          debugger;
          dispatch(getProductReccomendationItemsSuccess(response.data))
        })
        .catch((err) => {
          dispatch(getProductError(err))
        })
    }
  }

   return function(dispatch) {
    axios.get(productReccomendationApiUrl)
      .then((response) => {
        return response.data;
      }).then((response) =>{
        let ids = response.map( (item) => item.itemId);
        let idStr = ids.splice(0,10).join(',');
        productLookupApiURL = productConst.PRODUCT_LOOKUP_API_URL + '?ids=' + idStr + '&apiKey=' + productConst.PRODUCT_API_KEY;
        dispatch(getProductReccomendation(productLookupApiURL));
        dispatch(push(`/reccomendation/${itemId}`))
      })
      .catch((err) => {
        dispatch(getProductError(err))
      })
  }
    
};