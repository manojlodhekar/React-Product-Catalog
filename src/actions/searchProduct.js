import * as type from '../actions/actionTypes';
import productConst from '../const/productConst';
import axios from 'axios';

export function searchProduct(searchProductStr){
    
  let productApiURL = productConst.PRODUCT_SEARCH_API_URL + '?query='  + searchProductStr + '&format=json&apiKey=' + productConst.PRODUCT_API_KEY; ;  
  let productLookApi = ''; 

  const getProductSuccess = (response) => {
    debugger;
    console.log(response)
     return {
      type: type.SEARCH_PRODUCT,
      payload: response
    }
  }
  const getProductLookup = (url) => {
    return function(dispatch) {
      axios.get(url)
        .then((response) => {
          debugger;
          dispatch(getProductSuccess(response.data))
        })
        .catch((err) => {
          dispatch(getProductError(err))
        })
    }
  }
  

  const getProductError = (err) => {
    err = "";
    return {
      type: type.PRODUCT_API_ERROR,
      payload: true
    }
  }
  
   return function(dispatch) {
    axios.get(productApiURL)
      .then((response) => {
        return response.data;
      }).then((response) =>{
        let ids = response.items.map( (item) => item.itemId);
        let idStr = ids.splice(0,10).join(',');
        productLookApi = productConst.PRODUCT_LOOKUP_API_URL + '?ids=' + idStr + '&apiKey=' + productConst.PRODUCT_API_KEY;
        dispatch(getProductLookup(productLookApi))
      })
      .catch((err) => {
        dispatch(getProductError(err))
      })
  }
    
};
export function setApiError(isError){
  return {
    type: type.PRODUCT_API_ERROR,
    payload: false
  }
}