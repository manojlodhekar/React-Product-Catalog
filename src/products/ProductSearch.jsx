import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import PropTypes from 'prop-types';
import { Input, Button, Divider, Icon } from 'semantic-ui-react';
import { searchProduct } from '../actions/searchProduct';
import { getReccomendationProductList } from '../actions/getReccomendationProductList';
import productConst from '../const/productConst';
import ProductList from './ProductList';
import ProductHeader from './ProductHeader';
import  './ProductStyle.css'


class ProductSearch extends Component {
    constructor(props){
        super(props);
        this.state = {
            searchStr : ''
        }
    };

    componentWillMount(){
    }

    searchedStr = event =>{
        this.setState({
            searchStr: event.currentTarget.value
        })
    }
    handleKeyPress = event =>{
        if (event.key === 'Enter') {
            this.props.actions.searchProduct(this.state.searchStr);
          }
    }
    searchProductsByStr = event =>{
        this.props.actions.searchProduct(this.state.searchStr)
    }
    productSelection = (item) =>{
        if(item){
            this.props.actions.getReccomendationProductList(item.itemId);
        }
    }
    render(){
        return(
           <div className="container">
            <ProductHeader subHeaderText={productConst.PRODUCT_MAIN_PAGE_SUBHEADER}/>
                
                <div className="searchedContent">

                 <Input icon placeholder={productConst.PLACEHOLDER_TEXT}
                 id='searchTxt'
                 value={this.state.searchStr}
                 onChange={e => this.searchedStr(e)}
                 onKeyPress={e => this.handleKeyPress(e)}
                 className="searchInput">
                    <input />
                    <Icon name='search' />
                    </Input>

                    <Button id="SearchBtn" onClick={e => this.searchProductsByStr(e)}>Search</Button>
                    </div>
                    <Divider />
               <ProductList items={this.props.productList} 
               linkUrl="/reccomendation/"
               productSelection={(item) => this.productSelection(item)}/>
            </div>
        );
    };
}

ProductSearch.propTypes = {
    productList: PropTypes.array.isRequired,
    actions: PropTypes.object.isRequired
}

function mapStateToProps(state){
    return{
        productList: state.products.productList
    };
}

function mapDispatchToProps(dispatch){
    return {
        actions: bindActionCreators( { searchProduct, getReccomendationProductList }, dispatch),
    };
}

export default connect(mapStateToProps, mapDispatchToProps)(ProductSearch);
