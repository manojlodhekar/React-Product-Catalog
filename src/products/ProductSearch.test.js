import React, { Component } from 'react';
import ProductList from './ProductList';
import ProductHeader from './ProductHeader';
import { Input } from 'semantic-ui-react';
import * as actions from '../actions/getProductDetails';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import moxios from 'moxios';
import initialState from '../store/initialState'
import ProductSearch from './ProductSearch';
import { mapStateToProps, mapDispatchToProps } from './ProductSearch';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { shallow, mount } from 'enzyme';



Enzyme.configure({ adapter: new Adapter() });
jest.mock('react-dom');
//const initialState = {}; 

const mockStore = configureStore();
let store;

describe('Product Search Component', () => {
    let productSearch;
    const mockSearchProductfn = jest.fn();
    const mockGetReccomendationProductListFn = jest.fn();
    
     beforeEach(() => {
        store = mockStore(initialState)
       productSearch = mount(
        <ProductSearch store={store} searchProduct={mockSearchProductfn} getReccomendationProductList={mockGetReccomendationProductListFn}>
          <ProductHeader />
          <ProductList />
          <Input />
        </ProductSearch>
      );
    })


     it('should render the container component', () => {
        
        const product = productSearch.find(ProductList);
        expect(productSearch.find(product)).toBeDefined();
      });

      it('should render the productHeader component', () => {
        
        const productHeader = productSearch.find(ProductHeader);
        expect(productSearch.find(productHeader)).toBeDefined();
      });

     it('Snapshot of Product Search', () => {
        expect(productSearch).toMatchSnapshot();
    });

    //  it('should call the mock search Product function', () => {
    //     productSearch.find('#SearchBtn').simulate(
    //       'onClick', 
    //       {preventDefault() {}}
    //     )
    //     //expect(mockLoginfn.mock.calls.length).toBe(1)
    //    });


      //  it('should be called with the email and password in the state as arguments', () => {
        
      //   let searchInput = productSearch.find(Input);
      //   console.log(searchInput);
      //   productSearch.find('#searchTxt').simulate(
      //     'change', 
      //     {target: 
      //       {value: 'watch'}
      //     }
      //   )
        
      //   // simulate form submission   
      //   productSearch.find('#SearchBtn').simulate(
      //     'onClick', 
      //     {preventDefault() {}}
      //   )
      //   //expect(mockSearchProductfn.mock.calls[1][0]).toEqual('watch')
      //  })

 });
