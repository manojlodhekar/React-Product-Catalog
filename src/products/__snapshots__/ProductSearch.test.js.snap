// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Search Component Snapshot of Product Search 1`] = `
ReactWrapper {
  "length": 1,
  Symbol(enzyme.__unrendered__): <Connect(ProductSearch)
    getReccomendationProductList={[Function]}
    searchProduct={[Function]}
    store={
        Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          }
    }
>
    <ProductHeader />
    <ProductList />
    <Input
        type="text"
    />
</Connect(ProductSearch)>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__node__): null,
  Symbol(enzyme.__nodes__): Array [
    null,
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
